src/hotspot/share/opto/loopUnswitch.cpp

```diff
@@ -222,69 +227,172 @@ bool PhaseIdealLoop::has_control_dependencies_from_predicates(LoopNode* head) co
-  IfNode* iff = (unswitch_iff->Opcode() == Op_RangeCheck) ? new RangeCheckNode(entry, bol, unswitch_iff->_prob, unswitch_iff->_fcnt) :
-    new IfNode(entry, bol, unswitch_iff->_prob, unswitch_iff->_fcnt);
-  register_node(iff, outer_loop, entry, dom_depth(entry));
-  IfProjNode* iffast = new IfTrueNode(iff);
-  register_node(iffast, outer_loop, iff, dom_depth(iff));
-  IfProjNode* ifslow = new IfFalseNode(iff);
-  register_node(ifslow, outer_loop, iff, dom_depth(iff));
-
-  // Clone the loop body.  The clone becomes the slow loop.  The
-  // original pre-header will (illegally) have 3 control users
-  // (old & new loops & new if).
-  clone_loop(loop, old_new, dom_depth(head->skip_strip_mined()), mode, iff);
-  assert(old_new[head->_idx]->is_Loop(), "" );
-
-  // Fast (true) and Slow (false) control
-  IfProjNode* iffast_pred = iffast;
-  IfProjNode* ifslow_pred = ifslow;
-  clone_parse_and_assertion_predicates_to_unswitched_loop(loop, old_new, iffast_pred, ifslow_pred);
-
-  Node* l = head->skip_strip_mined();
-  _igvn.replace_input_of(l, LoopNode::EntryControl, iffast_pred);
-  set_idom(l, iffast_pred, dom_depth(l));
-  LoopNode* slow_l = old_new[head->_idx]->as_Loop()->skip_strip_mined();
-  _igvn.replace_input_of(slow_l, LoopNode::EntryControl, ifslow_pred);
-  set_idom(slow_l, ifslow_pred, dom_depth(l));
+#ifndef PRODUCT
+void PhaseIdealLoop::trace_loop_unswitching_impossible(const LoopNode* original_head) {
+  if (TraceLoopUnswitching) {
+    tty->print_cr("Loop Unswitching \"%d %s\" not possible due to control dependencies",
+                  original_head->_idx, original_head->Name());
+  }
+}
+
+void PhaseIdealLoop::trace_loop_unswitching_count(IdealLoopTree* loop, LoopNode* original_head) {
+  if (TraceLoopOpts) {
+    tty->print("Unswitch   %d ", original_head->unswitch_count() + 1);
+    loop->dump_head();
+  }
+}
 
-  recompute_dom_depth();
+void PhaseIdealLoop::trace_loop_unswitching_result(const UnswitchedLoopSelector& unswitched_loop_selector,
+                                                   const LoopNode* original_head, const LoopNode* new_head) {
+  if (TraceLoopUnswitching) {
+    IfNode* unswitch_candidate = unswitched_loop_selector.unswitch_candidate();
+    IfNode* loop_selector = unswitched_loop_selector.selector();
+    tty->print_cr("Loop Unswitching:");
+    tty->print_cr("- Unswitch-Candidate-If: %d %s", unswitch_candidate->_idx, unswitch_candidate->Name());
+    tty->print_cr("- Loop-Selector-If: %d %s", loop_selector->_idx, loop_selector->Name());
+    tty->print_cr("- True-Path-Loop (=Orig): %d %s", original_head->_idx, original_head->Name());
+    tty->print_cr("- False-Path-Loop (=Clone): %d %s", new_head->_idx, new_head->Name());
+  }
+}
+#endif
 
-  return iff;
+void PhaseIdealLoop::revert_to_normal_loop(const LoopNode* loop_head) {
+  if (loop_head->is_CountedLoop() && !loop_head->as_CountedLoop()->is_normal_loop()) {
+    loop_head->as_CountedLoop()->set_normal_loop();
+  }
```

Comment:
```suggestion
void PhaseIdealLoop::revert_to_normal_loop(const LoopNode* loop_head) {
  CountedLoopNode* cl = loop_head->isa_CountedLoop();
  if (cl != nullptr && !cl->is_normal_loop()) {
    cl->set_normal_loop();
  }
```

Reply:
Would make the code easier to read

Reply:
I just moved the code but I can clean it up as you suggested.